### PR TITLE
SAA-1306 changes to support recognition of prisoner permanently released to hospital on offender released event.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
It was previously thought RELEASE_TO_HOSPITAL on an offender released event was a temporary release ergo allocations and appointments were not ended/cancelled.  It turns out it is a permanent release from a given prison so allocations and appointments should be ended/cancelled accordingly.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReleasedEventHandler.kt [hunk 1]
```diff
@@ -80,17 +81,14 @@ class OffenderReleasedEventHandler(
...
-      fullInfo = true,
-      extraInfo = true,
-    ).block()
+    prisonSearchApiClient.findByPrisonerNumber(prisonerNumber = event.prisonerNumber())
+      ?: throw NullPointerException("Prisoner search lookup failed for prisoner ${event.prisonerNumber()}")
 
-  private fun InmateDetail.getDeallocationReasonForReleasedPrisoner(event: OffenderReleasedEvent) =
+  private fun getDeallocationReasonForReleasedPrisoner(prisoner: Prisoner, event: OffenderReleasedEvent) =
     when {
-      isReleasedOnDeath() -> DeallocationReason.DIED
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReleasedEventHandler.kt:91]: The logic for determining a deallocation reason has been relaxed quite a bit.

This should mean there is now far less chance of us not processing release events successfully.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReleasedEventHandler.kt [hunk 2]
```diff
@@ -80,17 +81,14 @@ class OffenderReleasedEventHandler(
...
-      isReleasedFromRemand() -> DeallocationReason.RELEASED
-      isReleasedFromCustodialSentence() -> DeallocationReason.RELEASED
-      else -> log.warn("Unable to determine release reason for prisoner ${event.prisonerNumber()}").let { null }
+      prisoner.isRestrictedPatient() -> RELEASED.also { log.info("Released restricted patient ${event.prisonerNumber()} from prison ${event.prisonCode()}") }
+      prisoner.isInactiveOut() -> RELEASED.also { log.info("Released inactive out prisoner ${event.prisonerNumber()} from prison ${event.prisonCode()}") }
+      prisoner.isAtDifferentLocationTo(event.prisonCode()) -> RELEASED.also { log.info("Released prisoner ${event.prisonerNumber()} from prison ${event.prisonCode()} now at ${prisoner.prisonId}") }
+      else -> TEMPORARILY_RELEASED.also { log.info("Temporary release or transfer of prisoner ${event.prisonerNumber()} from prison ${event.prisonCode()}") }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/events/handlers/OffenderReleasedEventHandler.kt:92]: If none of the above then we default to temporary release or transfer, we will no longer just log and ignore it.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/InboundEventsIntegrationTest.kt [hunk 3]
```diff
@@ -181,95 +192,22 @@ class InboundEventsIntegrationTest : IntegrationTestBase() {
...
-
   @Test
   @Sql("classpath:test_data/seed-offender-released-event-deletes-attendances.sql")
   fun `permanent release of prisoner removes any future attendances at time of event being raised`() {
+    prisonerSearchApiMockServer.stubSearchByPrisonerNumber(
+      PrisonerSearchPrisonerFixture.instance(
+        prisonerNumber = "A11111A",
+        inOutStatus = Prisoner.InOutStatus.OUT,
+        status = "INACTIVE OUT",
+        confirmedReleaseDate = null,
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/InboundEventsIntegrationTest.kt:203]: Are there any tests for those prisoners we saw mid-transfer yesterday? i.e. they have been booked out of one prison but not booked into the receiving prison, so appear as prison code TRN.
`"status": "INACTIVE TRN",
  "lastMovementTypeCode": "TRN",
  "lastMovementReasonCode": "APPEALS",
  "inOutStatus": "TRN",
  "prisonId": "TRN",
  "prisonName": "Transfer",`

**Review by timharrison-moj — APPROVED** [review]
LGTM. One question about whilst mid-transfer.

**Pull request merged**